Link error with OpenCV 2.4.13.2 and FFmpeg 1.2.12

##### System information (version)

- OpenCV => 2.4.13.2
- Operating System / Platform => Slackware Linux 14.0
- Compiler => GCC 4.7.1

##### Detailed description

Recently released OpenCV 2.4.13.2 fails to build on Slackware Linux 14.0 with ffmpeg 1.2.12 installed. The errors are:

```
Linking CXX executable ../../bin/opencv_perf_core
cd /usr/src/tmp/opencv-2.4.13.2/modules/core && /usr/bin/cmake -E cmake_link_script CMakeFiles/opencv_perf_core.dir/link.txt --verbose=1
/usr/bin/c++   -O2 -march=i486 -mtune=i686   -fsigned-char -W -Wall -Werror=return-type -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wno-narrowing -Wno-delete-non-virtual-dtor -Wno-comment -Wno-array-bounds -fdiagnostics-show-option -pthread -march=i686 -fno-omit-frame-pointer -msse -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mfpmath=sse -ffunction-sections -O2 -DNDEBUG  -DNDEBUG       CMakeFiles/opencv_perf_core.dir/perf/perf_main.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_convertTo.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_inRange.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_minmaxloc.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_arithm.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_merge.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_dft.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_dot.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_split.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_mat.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_addWeighted.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_math.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_abs.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_bitwise.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_reduce.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_compare.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_norm.cpp.o CMakeFiles/opencv_perf_core.dir/perf/perf_stat.cpp.o  -o ../../bin/opencv_perf_core -rdynamic ../../lib/libopencv_core.so.2.4.13 ../../lib/libopencv_ts.a ../../lib/libopencv_highgui.so.2.4.13 ../../lib/libopencv_core.so.2.4.13 ../../lib/libopencv_flann.so.2.4.13 ../../lib/libopencv_imgproc.so.2.4.13 ../../lib/libopencv_highgui.so.2.4.13 ../../lib/libopencv_features2d.so.2.4.13 ../../lib/libopencv_calib3d.so.2.4.13 ../../lib/libopencv_video.so.2.4.13 -ldl -lm -lpthread -lrt ../../lib/libopencv_features2d.so.2.4.13 ../../lib/libopencv_highgui.so.2.4.13 ../../lib/libopencv_flann.so.2.4.13 ../../lib/libopencv_imgproc.so.2.4.13 ../../lib/libopencv_core.so.2.4.13 -ldl -lm -lpthread -lrt -Wl,-rpath,/usr/src/tmp/opencv-2.4.13.2/lib
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_codec_get_id'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_seek_frame'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_network_init'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avpicture_get_size'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_find_stream_info'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_mallocz'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `sws_scale'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_guess_format'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_read_frame'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avio_close'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_guess_codec'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_freep'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_lockmgr_register'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_write_trailer'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_new_stream'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_rescale_q'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_encode_video2'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_close_input'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_align_dimensions2'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_close'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_alloc_frame'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_write_header'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_find_encoder'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_flush_buffers'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_write_frame'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_log_set_level'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_free'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_open2'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_free_frame'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `sws_getContext'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_dict_set'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_decode_video2'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_free_context'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_register_all'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avpicture_fill'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `sws_freeContext'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avio_open'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `sws_getCachedContext'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_init_packet'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_open_input'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_dict_free'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avcodec_find_decoder'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_malloc'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_sub_q'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `av_free_packet'
../../lib/libopencv_highgui.so.2.4.13: undefined reference to `avformat_alloc_context'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/opencv_perf_core] Error 1
make[2]: Leaving directory `/usr/src/tmp/opencv-2.4.13.2'
make[1]: *** [modules/core/CMakeFiles/opencv_perf_core.dir/all] Error 2
make[1]: Leaving directory `/usr/src/tmp/opencv-2.4.13.2'
make: *** [all] Error 2
```

##### Other information

Previous release 2.4.13 builds just fine.

```
  In file included from /usr/include/libavutil/avutil.h:238:0,
                   from /usr/include/libavutil/samplefmt.h:24,
                   from /usr/include/libavcodec/avcodec.h:30,
                   from /usr/include/libavformat/avformat.h:203,
                   from /usr/src/tmp/opencv-2.4.13.2/cmake/checks/ffmpeg_test.cpp:4:

  /usr/include/libavutil/common.h: In function 'int32_t av_clipl_int32_c(int64_t)':

  /usr/include/libavutil/common.h:175:47: error: 'UINT64_C' was not declared in this scope
```

Could you try to update CMake command:
`CXXFLAGS=-D__STDC_CONSTANT_MACROS cmake ...`
or
`cmake -DCMAKE_CXX_FLAGS=-D__STDC_CONSTANT_MACROS ...`

And probably do the same trick with `__STDC_LIMIT_MACROS`.


P.S. There is [StackOverflow answer](http://stackoverflow.com/a/986584) about these defines.

--- a/cmake/OpenCVModule.cmake
+++ b/cmake/OpenCVModule.cmake
@@ -170,6 +170,10 @@ macro(ocv_add_module _name)
       return() # extra protection from redefinition
     endif()
     project(${the_module})
+    add_definitions(
+        -D_USE_MATH_DEFINES  # M_PI constant in MSVS
+        -D__STDC_CONSTANT_MACROS -D__STDC_LIMIT_MACROS  # to use C libraries from C++ code (ffmpeg)
+    )
   endif(OPENCV_INITIAL_PASS)
 endmacro()
 

